app/main.py

```diff
@@ -1,214 +1,115 @@
+class Knight:
+    def __init__(self, name: str, power: int, hp: int, armour: str,
+                 weapon: str, potion: str):
+        self.name = name
+        self.base_power = power
+        self.hp = hp
+        self.armour = armour or []
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = self.calculate_protection()
+        self.power = self.calculate_power()
+
+    def calculate_protection(self):
+        protection = sum(a["protection"] for a in self.armour)
+        if self.potion and "protection" in self.potion.get("effect", {}):
+            protection += self.potion["effect"]["protection"]
+        return protection
+
+    def calculate_power(self):
+        power = self.base_power + (self.weapon["power"] if self.weapon else 0)
+        if self.potion and "power" in self.potion.get("effect", {}):
```

Comment:
Similar to the `calculate_protection` method, ensure that the `potion` attribute is always structured correctly to avoid issues when accessing 'effect'.